File: sdk-typescript/src/governance/signal-interface-v1.ts
import { createHash } from 'node:crypto';
import { canonicalJsonForHash } from '../policy/tool-pinning.js';
import { compareCodePoints } from './strict-canonical.js';

export const SIGNAL_DECLARATION_V1_SCHEMA = 'obsvr-signal-declaration-v1' as const;
export const SIGNAL_OBSERVATION_V1_SCHEMA = 'obsvr-signal-observation-v1' as const;
export const SIGNAL_RESOLUTION_V1_SCHEMA = 'obsvr-signal-resolution-v1' as const;
const HASH_RE = /^[0-9a-f]{64}$/;
export interface SignalDeclarationV1Input { signal_id: string; version: string; determinism: 'deterministic' | 'probabilistic'; locality: 'local' | 'remote'; timeout_ms: number; cache_ttl_ms: number; failure_disposition: 'deny' | 'defer' | 'ignore'; }
export interface SignalObservationV1Input { signal_id: string; version: string; input_hash: string; status: 'matched' | 'not_matched' | 'error' | 'timeout'; labels: string[]; score_bps?: number; provenance_hash: string; evaluated_at_ms: number; latency_ms: number; cache_state: 'hit' | 'miss' | 'not_cacheable'; }

export class SignalInterfaceV1ValidationError extends Error { constructor(message: string) { super(message); this.name = 'SignalInterfaceV1ValidationError'; } }
function fail(message: string): never { throw new SignalInterfaceV1ValidationError(message); }
function text(value: unknown, field: string): string { if (typeof value !== 'string' || !value.trim() || Buffer.byteLength(value.trim(), 'utf8') > 256) fail(`${field} must be nonblank and at most 256 UTF-8 bytes`); return value.trim(); }
function hash(value: unknown, field: string): string { if (typeof value !== 'string' || !HASH_RE.test(value)) fail(`${field} must be a lowercase SHA-256 hash`); return value; }
function integer(value: unknown, field: string, max = Number.MAX_SAFE_INTEGER): number { if (typeof value !== 'number' || !Number.isSafeInteger(value) || value < 0 || value > max) fail(`${field} must be a nonnegative safe integer no greater than ${max}`); return value; }

export function buildSignalDeclarationV1(input: SignalDeclarationV1Input) {
  if (!input || typeof input !== 'object' || Array.isArray(input)) fail('declaration must be an object'); const raw = input as unknown as Record<string, unknown>;
  const unknown = Object.keys(raw).filter((key) => !['schema', 'signal_id', 'version', 'determinism', 'locality', 'timeout_ms', 'cache_ttl_ms', 'failure_disposition'].includes(key)).sort(compareCodePoints); if (unknown.length) fail(`declaration contains unsupported field: ${unknown[0]}`);
  if (raw.schema !== undefined && raw.schema !== SIGNAL_DECLARATION_V1_SCHEMA) fail('declaration schema is invalid');
  if (!['deterministic', 'probabilistic'].includes(String(raw.determinism)) || !['local', 'remote'].includes(String(raw.locality)) || !['deny', 'defer', 'ignore'].includes(String(raw.failure_disposition))) fail('declaration enum is invalid');
  const timeout = integer(raw.timeout_ms, 'timeout_ms', 300_000); if (timeout === 0) fail('timeout_ms must be greater than zero');
  return { schema: SIGNAL_DECLARATION_V1_SCHEMA, signal_id: text(raw.signal_id, 'signal_id'), version: text(raw.version, 'version'), determinism: raw.determinism as SignalDeclarationV1Input['determinism'], locality: raw.locality as SignalDeclarationV1Input['locality'], timeout_ms: timeout, cache_ttl_ms: integer(raw.cache_ttl_ms, 'cache_ttl_ms', 86_400_000), failure_disposition: raw.failure_disposition as SignalDeclarationV1Input['failure_disposition'], authoritative_allow: false as const };
}

export function buildSignalObservationV1(input: SignalObservationV1Input) {
  if (!input || typeof input !== 'object' || Array.isArray(input)) fail('observation must be an object'); const raw = input as unknown as Record<string, unknown>;
  const unknown = Object.keys(raw).filter((key) => !['schema', 'signal_id', 'version', 'input_hash', 'status', 'labels', 'score_bps', 'provenance_hash', 'evaluated_at_ms', 'latency_ms', 'cache_state'].includes(key)).sort(compareCodePoints); if (unknown.length) fail(`observation contains unsupported field: ${unknown[0]}`);
  if (raw.schema !== undefined && raw.schema !== SIGNAL_OBSERVATION_V1_SCHEMA) fail('observation schema is invalid');
  if (!['matched', 'not_matched', 'error', 'timeout'].includes(String(raw.status)) || !['hit', 'miss', 'not_cacheable'].includes(String(raw.cache_state))) fail('observation enum is invalid');
  if (!Array.isArray(raw.labels) || raw.labels.length > 128) fail('labels must contain at most 128 items'); const labels = [...new Set(raw.labels.map((item, i) => text(item, `labels[${i}]`)))].sort(compareCodePoints);
  const result: Record<string, unknown> = { schema: SIGNAL_OBSERVATION_V1_SCHEMA, signal_id: text(raw.signal_id, 'signal_id'), version: text(raw.version, 'version'), input_hash: hash(raw.input_hash, 'input_hash'), status: raw.status, labels, provenance_hash: hash(raw.provenance_hash, 'provenance_hash'), evaluated_at_ms: integer(raw.evaluated_at_ms, 'evaluated_at_ms'), latency_ms: integer(raw.latency_ms, 'latency_ms'), cache_state: raw.cache_state };
  if (raw.score_bps !== undefined) result.score_bps = integer(raw.score_bps, 'score_bps', 10_000); return result;
}

export function resolveSignalV1(declarationInput: SignalDeclarationV1Input, observationInput: SignalObservationV1Input) {
  const declaration = buildSignalDeclarationV1(declarationInput); const observation = buildSignalObservationV1(observationInput);
  if (declaration.signal_id !== observation.signal_id || declaration.version !== observation.version) fail('observation does not match declaration');
  const failure = observation.status === 'error' || observation.status === 'timeout';
  const requiredOutcome = failure && declaration.failure_disposition !== 'ignore' ? declaration.failure_disposition.toUpperCase() : null;
  const body = { schema: SIGNAL_RESOLUTION_V1_SCHEMA, declaration, observation, fact: { matched: observation.status === 'matched', labels: observation.labels, ...(observation.score_bps !== undefined ? { score_bps: observation.score_bps } : {}) }, required_outcome: requiredOutcome, authoritative_allow: false as const };
  return { ...body, resolution_hash: createHash('sha256').update(`obsvr-signal-resolution/1\0${canonicalJsonForHash(body)}`, 'utf8').digest('hex') };
}

export function signalResolutionToOtelAttributesV1(resolution: ReturnType<typeof resolveSignalV1>) { return { 'obsvr.signal.schema': resolution.schema, 'obsvr.signal.id': resolution.declaration.signal_id, 'obsvr.signal.version': resolution.declaration.version, 'obsvr.signal.determinism': resolution.declaration.determinism, 'obsvr.signal.locality': resolution.declaration.locality, 'obsvr.signal.status': resolution.observation.status, 'obsvr.signal.provenance_hash': resolution.observation.provenance_hash, 'obsvr.signal.resolution_hash': resolution.resolution_hash, 'obsvr.signal.latency_ms': resolution.observation.latency_ms, 'obsvr.signal.authoritative_allow': false }; }
export function signalResolutionToOpaInputV1(resolution: ReturnType<typeof resolveSignalV1>) { return { obsvr_signal: { id: resolution.declaration.signal_id, version: resolution.declaration.version, matched: resolution.fact.matched, labels: resolution.fact.labels, score_bps: resolution.fact.score_bps ?? null, provenance_hash: resolution.observation.provenance_hash, required_outcome: resolution.required_outcome, authoritative_allow: false, resolution_hash: resolution.resolution_hash } }; }
export function signalResolutionToCedarContextV1(resolution: ReturnType<typeof resolveSignalV1>) { return { obsvrSignalId: resolution.declaration.signal_id, obsvrSignalVersion: resolution.declaration.version, obsvrSignalMatched: resolution.fact.matched, obsvrSignalLabels: resolution.fact.labels, obsvrSignalScoreBps: resolution.fact.score_bps ?? 0, obsvrSignalProvenanceHash: resolution.observation.provenance_hash, obsvrSignalRequiredOutcome: resolution.required_outcome ?? '', obsvrSignalAuthoritativeAllow: false, obsvrSignalResolutionHash: resolution.resolution_hash }; }

File: sdk-python/obsvr/signal_interface_v1.py
"""Bounded evaluator signals; the deterministic local kernel retains authority."""

from __future__ import annotations

import hashlib
from typing import Any, Dict

from .strict_canonical import code_point_key
from .tool_pinning import _canonical_json_for_hash

SIGNAL_DECLARATION_V1_SCHEMA = "obsvr-signal-declaration-v1"
SIGNAL_OBSERVATION_V1_SCHEMA = "obsvr-signal-observation-v1"
SIGNAL_RESOLUTION_V1_SCHEMA = "obsvr-signal-resolution-v1"
_HEX = frozenset("0123456789abcdef")


class SignalInterfaceV1ValidationError(ValueError):
    pass


def _fail(message: str) -> None:
    raise SignalInterfaceV1ValidationError(message)


def _text(value: Any, field: str) -> str:
    if not isinstance(value, str) or not value.strip() or len(value.strip().encode()) > 256:
        _fail(f"{field} must be nonblank and at most 256 UTF-8 bytes")
    return value.strip()


def _hash(value: Any, field: str) -> str:
    if not isinstance(value, str) or len(value) != 64 or any(c not in _HEX for c in value):
        _fail(f"{field} must be a lowercase SHA-256 hash")
    return value


def _integer(value: Any, field: str, maximum: int = 9_007_199_254_740_991) -> int:
    if isinstance(value, bool) or not isinstance(value, int) or not 0 <= value <= maximum:
        _fail(f"{field} must be a nonnegative safe integer no greater than {maximum}")
    return value


def build_signal_declaration_v1(input_value: Dict[str, Any]) -> Dict[str, Any]:
    if not isinstance(input_value, dict):
        _fail("declaration must be an object")
    unknown = sorted(set(input_value) - {"schema", "signal_id", "version", "determinism", "locality", "timeout_ms", "cache_ttl_ms", "failure_disposition"}, key=code_point_key)
    if unknown:
        _fail(f"declaration contains unsupported field: {unknown[0]}")
    if "schema" in input_value and input_value["schema"] != SIGNAL_DECLARATION_V1_SCHEMA:
        _fail("declaration schema is invalid")
    if input_value.get("determinism") not in {"deterministic", "probabilistic"} or input_value.get("locality") not in {"local", "remote"} or input_value.get("failure_disposition") not in {"deny", "defer", "ignore"}:
        _fail("declaration enum is invalid")
    timeout = _integer(input_value.get("timeout_ms"), "timeout_ms", 300_000)
    if timeout == 0:
        _fail("timeout_ms must be greater than zero")
    return {"schema": SIGNAL_DECLARATION_V1_SCHEMA, "signal_id": _text(input_value.get("signal_id"), "signal_id"), "version": _text(input_value.get("version"), "version"), "determinism": input_value["determinism"], "locality": input_value["locality"], "timeout_ms": timeout, "cache_ttl_ms": _integer(input_value.get("cache_ttl_ms"), "cache_ttl_ms", 86_400_000), "failure_disposition": input_value["failure_disposition"], "authoritative_allow": False}


def build_signal_observation_v1(input_value: Dict[str, Any]) -> Dict[str, Any]:
    if not isinstance(input_value, dict):
        _fail("observation must be an object")
    unknown = sorted(set(input_value) - {"schema", "signal_id", "version", "input_hash", "status", "labels", "score_bps", "provenance_hash", "evaluated_at_ms", "latency_ms", "cache_state"}, key=code_point_key)
    if unknown:
        _fail(f"observation contains unsupported field: {unknown[0]}")
    if "schema" in input_value and input_value["schema"] != SIGNAL_OBSERVATION_V1_SCHEMA:
        _fail("observation schema is invalid")
    if input_value.get("status") not in {"matched", "not_matched", "error", "timeout"} or input_value.get("cache_state") not in {"hit", "miss", "not_cacheable"}:
        _fail("observation enum is invalid")
    raw_labels = input_value.get("labels")
    if not isinstance(raw_labels, list) or len(raw_labels) > 128:
        _fail("labels must contain at most 128 items")
    result = {"schema": SIGNAL_OBSERVATION_V1_SCHEMA, "signal_id": _text(input_value.get("signal_id"), "signal_id"), "version": _text(input_value.get("version"), "version"), "input_hash": _hash(input_value.get("input_hash"), "input_hash"), "status": input_value["status"], "labels": sorted({_text(item, f"labels[{i}]") for i, item in enumerate(raw_labels)}, key=code_point_key), "provenance_hash": _hash(input_value.get("provenance_hash"), "provenance_hash"), "evaluated_at_ms": _integer(input_value.get("evaluated_at_ms"), "evaluated_at_ms"), "latency_ms": _integer(input_value.get("latency_ms"), "latency_ms"), "cache_state": input_value["cache_state"]}
    if "score_bps" in input_value:
        result["score_bps"] = _integer(input_value["score_bps"], "score_bps", 10_000)
    return result


def resolve_signal_v1(declaration_input: Dict[str, Any], observation_input: Dict[str, Any]) -> Dict[str, Any]:
    declaration = build_signal_declaration_v1(declaration_input)
    observation = build_signal_observation_v1(observation_input)
    if declaration["signal_id"] != observation["signal_id"] or declaration["version"] != observation["version"]:
        _fail("observation does not match declaration")
    failure = observation["status"] in {"error", "timeout"}
    required = declaration["failure_disposition"].upper() if failure and declaration["failure_disposition"] != "ignore" else None
    fact = {"matched": observation["status"] == "matched", "labels": observation["labels"]}
    if "score_bps" in observation:
        fact["score_bps"] = observation["score_bps"]
    body = {"schema": SIGNAL_RESOLUTION_V1_SCHEMA, "declaration": declaration, "observation": observation, "fact": fact, "required_outcome": required, "authoritative_allow": False}
    body["resolution_hash"] = hashlib.sha256(f"obsvr-signal-resolution/1\0{_canonical_json_for_hash(body)}".encode()).hexdigest()
    return body


def signal_resolution_to_otel_attributes_v1(resolution: Dict[str, Any]) -> Dict[str, Any]:
    return {"obsvr.signal.schema": resolution["schema"], "obsvr.signal.id": resolution["declaration"]["signal_id"], "obsvr.signal.version": resolution["declaration"]["version"], "obsvr.signal.determinism": resolution["declaration"]["determinism"], "obsvr.signal.locality": resolution["declaration"]["locality"], "obsvr.signal.status": resolution["observation"]["status"], "obsvr.signal.provenance_hash": resolution["observation"]["provenance_hash"], "obsvr.signal.resolution_hash": resolution["resolution_hash"], "obsvr.signal.latency_ms": resolution["observation"]["latency_ms"], "obsvr.signal.authoritative_allow": False}


def signal_resolution_to_opa_input_v1(resolution: Dict[str, Any]) -> Dict[str, Any]:
    return {"obsvr_signal": {"id": resolution["declaration"]["signal_id"], "version": resolution["declaration"]["version"], "matched": resolution["fact"]["matched"], "labels": resolution["fact"]["labels"], "score_bps": resolution["fact"].get("score_bps"), "provenance_hash": resolution["observation"]["provenance_hash"], "required_outcome": resolution["required_outcome"], "authoritative_allow": False, "resolution_hash": resolution["resolution_hash"]}}


def signal_resolution_to_cedar_context_v1(resolution: Dict[str, Any]) -> Dict[str, Any]:
    return {"obsvrSignalId": resolution["declaration"]["signal_id"], "obsvrSignalVersion": resolution["declaration"]["version"], "obsvrSignalMatched": resolution["fact"]["matched"], "obsvrSignalLabels": resolution["fact"]["labels"], "obsvrSignalScoreBps": resolution["fact"].get("score_bps", 0), "obsvrSignalProvenanceHash": resolution["observation"]["provenance_hash"], "obsvrSignalRequiredOutcome": resolution["required_outcome"] or "", "obsvrSignalAuthoritativeAllow": False, "obsvrSignalResolutionHash": resolution["resolution_hash"]}
